x-pack/platform/packages/shared/kbn-streamlang/src/validation/validate_streamlang.ts

```diff
@@ -0,0 +1,753 @@
+    // Collect fields used as input by this processor
+    switch (step.action) {
+      case 'date':
+      case 'convert':
+      case 'replace':
+      case 'remove':
+      case 'grok':
+      case 'dissect':
+        if (step.from) fieldsUsed.push(step.from);
+        break;
+      case 'rename':
+        if (step.from) fieldsUsed.push(step.from);
+        break;
+      case 'set':
+        if (step.copy_from) fieldsUsed.push(step.copy_from);
+        break;
+      case 'append':
+      case 'drop_document':
+      case 'manual_ingest_pipeline':
+      case 'remove_by_prefix':
+        // These processors don't use specific fields in a way that requires type validation
+        break;
+      default: {
+        const _exhaustiveCheck: never = step;
+        return _exhaustiveCheck;
+      }
+    }
+
+    // Validate each field usage against current types
+    for (const field of fieldsUsed) {
+      const expectedTypes = getExpectedInputType(step, field);
+      if (expectedTypes) {
+        const actualType = fieldTypes.get(field);
+        if (actualType && actualType !== 'unknown' && !expectedTypes.includes(actualType)) {
+          errors.push({
+            type: 'type_mismatch',
+            message: i18n.translate('xpack.streamlang.validation.typeMismatchMessage', {
+              defaultMessage:
+                'Processor #{processorNumber} ({processorAction}) expects field "{fieldName}" to be of type {expectedTypes}, but it has type {actualType}',
+              values: {
+                processorNumber: i + 1,
+                processorAction: step.action,
+                fieldName: field,
+                expectedTypes: expectedTypes.join(' or '),
+                actualType,
+              },
+            }),
+            processorId,
+            field,
+            expectedType: expectedTypes.length === 1 ? expectedTypes[0] : expectedTypes,
+            actualType,
+          });
+        }
+      }
+    }
+
+    // Check for duplicate fields with different types in grok patterns
+    if (step.action === 'grok' && step.patterns) {
+      const patterns = Array.isArray(step.patterns) ? step.patterns : [step.patterns];
+      for (const pattern of patterns) {
+        // Extract all occurrences of each field in this pattern with their types
+        const fieldOccurrences = new Map<string, Set<string>>();
+        const regex = /%\{([^:}]+):([^:}]+)(?::([^}]+))?\}/g;
+        let match: RegExpExecArray | null;
+
+        while ((match = regex.exec(pattern)) !== null) {
+          const fieldName = match[2]?.trim();
+          const typeModifier = match[3]?.trim().toLowerCase();
+
+          if (!fieldName) continue;
+
+          // Determine the type for this occurrence
+          let occurrenceType: string;
+          if (typeModifier) {
+            if (['int', 'long', 'float', 'double'].includes(typeModifier)) {
+              occurrenceType = 'number';
+            } else if (['boolean', 'bool'].includes(typeModifier)) {
```

Comment:
You should [only need boolean](https://www.elastic.co/docs/explore-analyze/scripting/grok) (not bool).